### DISCUSSION TITLE
How to use PrivateCookieJar in middleware

### DISCUSSION THREAD

**Discussion by ghost** [tokio-rs/axum#2171] [Q&A] [answered]
I have been trying to use PrivateCookieJar to manage cookie but wasn't able to get it compile.

My middleware looks like this

```rust
pub async fn cookie_middleware<B>(
    cookie_jar: PrivateCookieJar,
    mut request: Request<B>,
    next: Next<B>,
) -> impl IntoResponse {
```

I am adding as a layer like this.

```
.layer(ServiceBuilder::new().layer(axum::middleware::from_fn(cookie_middleware)));
```

I got this error

```rust
^ the trait `Service<axum::http::Request<_>>` is not implemented for 
`axum::middleware::FromFn<fn(axum_extra::extract::cookie::private::PrivateCookieJar, axum::http::Request<_>, Next<_>) -> 
impl std::future::Future<Output = impl IntoResponse> {cookie_middleware::<_>}, (), Route, _>`
```

It compiles again when I remove `cookie_jar: PrivateCookieJar`

**Comment by ghost** [comment, marked as answer]
I managed to get it compiled after adding an `app_state` and do `impl FromRef<AppState> for Key`

`debug_middleware` might be helpful in this case. https://github.com/tokio-rs/axum/pull/1993

**Reply by davidpdrsn** [reply]
Relevant section in docs: https://docs.rs/axum-extra/0.7.7/axum_extra/extract/cookie/struct.PrivateCookieJar.html#example
